src/lib/core/base/base.ts

```diff
@@ -109,9 +109,21 @@ export abstract class BaseDirective implements OnDestroy, OnChanges {
     return this._elementRef.nativeElement;
   }
 
+  /** Add styles to the element using predefined style builder */
   protected addStyles(input: string, parent?: Object) {
-    const styles: StyleDefinition = this._styleBuilder!.buildStyles(input, parent);
-    this._applyStyleToElement(styles);
+    let styleDefinition: StyleDefinition;
+    if (this._styleCache.has(input)) {
+      styleDefinition = this._styleCache.get(input)!;
+      this._applyStyleToElement(styleDefinition);
+    } else {
+      const {shouldCache, styles} = this._styleBuilder!.buildStyles(input, parent);
+      this._applyStyleToElement(styles);
+      if (shouldCache) {
+        this._styleCache.set(input, styles);
+      }
+      styleDefinition = styles;
+    }
+    this._styleBuilder!.sideEffect(input, styleDefinition, parent);
```

Comment:
Why are you using the `!` here ?

Reply:
The style builder is an opt-in on the `BaseDirective`, and therefore can be `undefined`. But, to consciously use `addStyles` on `BaseDirective` implies that you are using a style builder, so the `!` means "trust me, it's there"